infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/CorfuLogReplicationRuntime.java

```diff
@@ -0,0 +1,295 @@
+ *                                                                                      |      |
+ *                                                                                      +------+
+ *            +---------------+     ALL STATES
+ *            |               |                                                     ON_CONNECTION_DOWN
+ *            | UNRECOVERABLE <---- ON_ERROR                                           (NON-LEADER)
+ *            |    STATE      |
+ *            |               |
+ *            +---------------+
+ *
+ *
+ * States:
+ * ------
+ *
+ * - WAITING_FOR_CONNECTIVITY    :: initial state, waiting for any connection to remote cluster to be established.
+ * - VERIFYING_REMOTE_LEADER     :: verifying the leader endpoint on remote cluster (querying all connected nodes)
+ * - NEGOTIATING                 :: negotiating against the leader endpoint
+ * - REPLICATING                 :: replicating data to remote cluster through the leader endpoint
+ * - STOPPED                     :: stop state machine, no error, just lost leadership so replication stops from this node
+ * - UNRECOVERABLE_STATE         :: error state, unrecoverable error reported by replication, transport or cluster manager, despite
+ *                                  being the leader node.
+ *
+ *
+ * Events / Transitions:
+ * --------------------
+ *
+ * - ON_CONNECTION_UP           :: connection to a remote endpoint comes UP
+ * - ON_CONNECTION_DOWN         :: connection to a remote endpoint comes DOWN
+ * - REMOTE_LEADER_NOT_FOUND,   :: remote leader not found
+ * - REMOTE_LEADER_FOUND,       :: remote leader found
+ * - REMOTE_LEADER_LOSS,        :: remote Leader Lost (remote node reports it is no longer the leader)
+ * - LOCAL_LEADER_LOSS          :: local node looses leadership
+ * - NEGOTIATION_COMPLETE,      :: negotiation succeeded and completed
+ * - NEGOTIATION_FAILED,        :: negotiation failed
+ * - STOPPED                    :: stop log replication server (fatal state)
+ *
+ * @author amartinezman
+ *
+ */
+@Slf4j
+public class CorfuLogReplicationRuntime {
+
+    // TODO(Gabriela): add site_flip (cluster_role_flip) event... probably we need a new state called finishing_ongoing_replication... and go to to stopped...
+
+    public static final int DEFAULT_TIMEOUT = 5000;
+
+    /**
+     * Current state of the FSM.
+     */
+    private volatile LogReplicationRuntimeState state;
+
+    /**
+     * Map of all Log Replication Communication FSM States (reuse single instance for each state)
+     */
+    @Getter
+    private Map<LogReplicationRuntimeStateType, LogReplicationRuntimeState> states = new HashMap<>();
+
+    /**
+     * Executor service for FSM state tasks
+     */
+    private ExecutorService communicationFSMWorkers;
+
+    /**
+     * Executor service for FSM event queue consume
+     */
+    private ExecutorService communicationFSMConsumer;
```

Comment:
Why do we need two thread pools, communicationFSMWorks and communicationFSMConsumer? Those tasks are really small, the overhead of context switch is bigger than running the task. Also, it  adds complexity to the system to make sure these two threads are sync.

Reply:
These two threads do not need to be in sync... I see it this way, imagine this scenario, you are doing negotiation and blocked waiting for the response to come back, but right when you start it, the connection comes down. If you don't have a consumer thread which is dedicated you are not able to process these async events.. so you just block until it timeout, while you could've abort and released resources and go ahead and re-connect and look for a new leader.

Reply:
It is cleaner that the first event is timeout and retires, and then process the second event. Processing two events at the same time will lead to an un-deterministic result due to timing issues.